Adding sequential_read

eyou102 opened the pull request:
Include methods allowing for reading file (or a chunk of a file) sequentially, looking for certain records to return as a dictionary
includes all supporting methods for splitting ping records by sector and translating fields

ericgyounkin committed:
adding read/decode methods

Adding methods for supporting reading/decoding the file sequentially, without mapping out all the byte locations, which is faster for chunking files
See read_first_datagram for an example

Fix read_EMdgmIIP  read_EMdgmIOP skipping unknown fields, have to use overall datagram size to determine the remainder

ericgyounkin committed:
more sequential_read stuff

translate sequential read output to standard field names
read attitude time from records instead of header
implement more efficient listofdicts2dictoflists
- in testing, found it improved from around 5ms to 1ms to run
add remainder of sequential_read_records needed utilities from par module

ericgyounkin committed:
fix sequential read - attitude/navigation

bug with kmall sequential read attitude, copy first list to avoid appending to data
concat att/nav blobs, resolve header vs data time

valschmidt commented on kmall.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+`#!/usr/bin/env python
```
Is the back quote at the beginning of this line intentional? What is this?

eyou102 replied:
Oops, not intentional.  Should I redo the pull request with changes?

rolker replied:
I think the pull request might get updated if you simply commit the fix again.

eyou102 replied:
OK, i see the change in the stack now.  Maybe that worked?

ericgyounkin committed:
fix bug with env path

valschmidt commented on kmall.py:
```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
+        recs_to_read = self._finalize_records(recs_to_read)
+        return recs_to_read
+
+    def translate_yawpitch(self, arr):
```
In this function, and the others that do this kind of translation, I think it would be good to name them something like translate_yawpitchstab_tonumeric(), so it is a bit more descriptive. Also in the doc strings for these, it would be good to try to briefly capture why one would want to do the translation.

eyou102 replied:
done.

valschmidt commented on kmall.py:
```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
+                totalrecs.append(split_rec)
+            return totalrecs
+
+    def _pad_to_dense(self, arr, padval=999.0, maxlen=500, override_type=None, detectioninfo=False):
```
Maybe add in the doc string why one would want to "pad to dense". If one doesn't know what you're up to, it be hard to understand the purpose of the function. [I think it's because you need a rectangular array of records for down-stream processing, but I'm not 100% sure myself.]

eyou102 replied:
done

valschmidt commented on kmall.py:
```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
+            totalrecs = []
+            pingtime = rec['header']['dgtime']
+            for sec in rec['txSectorInfo']['txSectorNumb']:
+                split_rec = copy.copy(rec)
```
Do you think it is better to use `copy` or `deepcopy` here? Certainly less memory intensive to use copy. But copy is kind of counter intuitive since changes to the copy change the original. If you keep `copy`, maybe make a note in the doc string that the new records contain references to the original record so there's some hope that I won't do something stupid and get something unexpected.

eyou102 replied:
A shallow copy is what we want I think.  I had an issue with changing the time stamp that led me to deepcopy at first, but the whole process takes 10x longer with deepcopy (my test went from 2 secs to about 20 secs)

ericgyounkin committed:
install/runtime params translation and other fixes

Add more documentation
Translate installation parameters string when using read_datagram (default otherwise is to not translate so as not to break existing systems)
Translate runtime parameters string when using read_datagram (default otherwise is to not translate so as not to break existing systems)

valschmidt commented on kmall.py:
```diff
@@ -3507,8 +3542,16 @@ def _divide_rec(self, rec):
         """
-        Appends the minimal required amount of zeroes at the end of each
-        array in the jagged array `M`, such that `M` looses its jaggedness.
+        Appends the minimal required amount of zeroes at the end of each array in the jagged array `M`, such that `M`
```
Eric, I was thinking about this method - I *think* that unlike the `.all` format, the `.kmall` format returns the same number of sounding records per MRZ datagram regardless of the number of detections for a given configuration. For example, for high-density mode with extra detection turned off, each MRZ record I think always 800 soundings, regardless of the number of valid detects. Invalid soundings are marked with a flag field. So it might be that you will never need to pad the data to be rectangular.  I will admit, I'm not 100% sure about this, but I think that's what I've seen.

eyou102 replied:
the padding isn't related to valid/invalid detections, its how I handle the ping record as a whole.  I take all the unique combinations of serial number/sector number/frequency, and divide the ping record by those indices.  So one MRZ datagram might end up being 3 ping records for me, see kmall._divide_rec.

As an example, for the file you provided (0007_20190513_154724_ASVBEN) I end up with the following ping records
(ping_systemid_sectornum_frequency)
ping_40_0_260000
ping_40_0_290000
ping_40_1_320000
ping_40_1_350000
ping_40_2_260000
ping_40_2_290000

each of these being stored independently.  It's admittedly weird, but it allows me to use the xarray Dataset object which requires the timestamps be unique, which in turn lets me to all the easy interpolation, etc.  I've found you have to isolate for serial number/sector number/frequency in order to avoid duplicate times in the record.  

Theoretically, the code I have will support non-sector based systems as well, it will just store as a single Dataset object.  If we move to include other file formats.

I used detection info later when I export to csv or to the sounding Dataset to flag soundings.

valschmidt commented on kmall.py:
```diff
@@ -15,9 +14,55 @@
 from pyproj import Proj
 from scipy import stats
 
+recs_categories = {'SKM': ['header.dgtime', 'sample.KMdefault.roll_deg', 'sample.KMdefault.pitch_deg',
```
Eric, what do think about moving these into the class? They could still be accessed outside the class by creating a class instance and then referencing them with `self.recs_categories`. 

I'm not sure they're meaningful outside the class so having them part of the class might clean things up a bit.

eyou102 replied:
Sure, no problem there.  I was trying to avoid muddying up your code,  Wasn't sure how you'd feel about 60 lines of dictionary in your init.  I think that would bother some people...

valschmidt commented on kmall.py:
```diff
@@ -3680,6 +3763,171 @@ def translate_mode_two(self, arr):
 
         return rslt
 
+    def translate_runtime_parameters_todict(self, r_text):
```
I REALLY like the way you've implemented this. I was playing around with it this afternoon and you've a really nice job of handling the myriad of runtime and installation parameters. Thanks Eric.

eyou102 replied:
Sure!  I had the same issue with par module, you really need a dict to make use of it.  And it allows for using the json library to dump to string and reload, which is handy for saving to file.  I have to do that to store it as attribute.

When you use xarray_conversion in kluster, you get a Dataset with the timestamped parameter entries as attributes which is super handy for figuring out when things changed, like this:

Attributes:
    _compute_beam_vectors_complete:        Thu Sep 17 19:58:14 2020
    _compute_orientation_complete:            Thu Sep 17 19:58:08 2020
    _conversion_complete:                            Thu Sep 17 19:58:04 2020
    _georeference_soundings_complete:      Thu Sep 17 19:58:57 2020
    _sound_velocity_correct_complete:         Thu Sep 17 19:58:23 2020
    installsettings_1557762444:                    {"operator_controller_version": "Empty...
    multibeam_files:                                      {'0007_20190513_154724_ASVBEN.kmall': ...
    output_path:                                           C:\collab\dasktest\data_dir\kmall_test...
    profile_1557757460:                               [[0.0, 1416.9000244140625], [0.1500000...
    runtimesettings_1557762444:                "\nSector coverage\nMax angle Port: ...
    secondary_system_serial_number:         [0]
    sector_identifier:                                    40_0_260000
    svmode:                                                 nearest in time
    system_serial_number:                          [53011]
    vertical_reference:                                 waterline
    xyz_crs:                                                  26917
    xyzrph:                                                  {'antenna_x': {'1557762443': '0.000'},...

valschmidt reviewed (commented):
Eric, this looks great. I'm going to merge it - there's just a few questions.

valschmidt commented:
I see your point. 
What if we define them in a a method and call this in the init. I think you’re point about mucking up the init is a good one. Then it would clean up the top of the code and the init both. 

> On Sep 18, 2020, at 4:41 PM, eyou102 <notifications@github.com> wrote:
> 
> 
> @eyou102 commented on this pull request.
> 
> In kmall.py <https://github.com/valschmidt/kmall/pull/7#discussion_r491180260>:
> 
> >  from pyproj import Proj
>  from scipy import stats
>  
> +recs_categories = {'SKM': ['header.dgtime', 'sample.KMdefault.roll_deg', 'sample.KMdefault.pitch_deg',
> Sure, no problem there. I was trying to avoid muddying up your code, Wasn't sure how you'd feel about 60 lines of dictionary in your init. I think that would bother some people...
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/valschmidt/kmall/pull/7#discussion_r491180260>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AASPJV2AII4IQUAUUAYBPRTSGPAV7ANCNFSM4ROZBYSQ>.
> 

------------------------------------------------------
Val Schmidt
Center for Coastal and Ocean Mapping / Joint Hydrographic Center
University of New Hampshire
Chase Ocean Engineering Lab
24 Colovos Road
Durham, NH 03824
e: vschmidt [AT] ccom.unh.edu
m: 614.286.3726

ghost commented:
Yeah sounds reasonable.  I can attack that on Monday.

ericgyounkin committed:
move recs_categories dicts to within kmall
